src/coprocessor/endpoint.rs

```diff
@@ -62,8 +65,76 @@ const OUTDATED_ERROR_MSG: &str = "request outdated.";
 
 const ENDPOINT_IS_BUSY: &str = "endpoint is busy";
 
+struct CopContextInner {
+    exec_local_metrics: ExecLocalMetrics,
+    basic_local_metrics: BasicLocalMetrics,
+    timer: Instant,
+    timeout: Duration,
+}
+
+impl CopContextInner {
+    fn collect(&mut self, region_id: u64, scan_tag: &str, metrics: ExecutorMetrics) {
+        self.exec_local_metrics
+            .collect(scan_tag, region_id, metrics);
+
+        let new_timer = Instant::now_coarse();
+        if new_timer.duration_since(self.timer) >= self.timeout {
+            self.exec_local_metrics.flush();
+            self.basic_local_metrics.flush();
+            self.timer = new_timer;
+        }
+    }
+}
+
+struct CopContext(RefCell<CopContextInner>);
+
+impl CopContext {
+    fn new(pd_task_sender: FutureScheduler<PdTask>) -> Self {
+        let inner = CopContextInner {
+            exec_local_metrics: ExecLocalMetrics::new(pd_task_sender),
+            basic_local_metrics: BasicLocalMetrics::default(),
+            timer: Instant::now_coarse(),
+            timeout: Duration::from_secs(1),
+        };
+        CopContext(RefCell::new(inner))
+    }
+}
+
+unsafe impl Sync for CopContext {}
```

Comment:
For similar reasons in https://github.com/pingcap/tikv/pull/2725, I think it is not good to impl Sync here since it is not Sync at all. However because it will be replaced by Future Pool very soon, I accept this implementation for now.